source/common/config/config_provider_impl.cc

```diff
@@ -0,0 +1,116 @@
+  }
+}
+
+bool ConfigSubscriptionInstanceBase::checkAndApplyConfig(const Protobuf::Message& config_proto,
+                                                         const std::string& config_name,
+                                                         const std::string& version_info) {
+  const uint64_t new_hash = MessageUtil::hash(config_proto);
+  if (config_info_ && config_info_.value().last_config_hash_ == new_hash) {
+    return false;
+  }
+
+  config_info_ = {new_hash, version_info};
+  ENVOY_LOG(debug, "{}: loading new configuration: config_name={} hash={}", name_, config_name,
+            new_hash);
+
+  ASSERT(!mutable_config_providers_.empty());
+  ConfigProvider::ConfigConstSharedPtr new_config;
+  for (auto* provider : mutable_config_providers_) {
+    // All bound mutable config providers must be of the same type (see the ASSERT... in
+    // bindConfigProvider()).
+    // This makes it safe to call any of the provider's onConfigProtoUpdate() to get a new config
+    // impl, which can then be passed to all providers.
+    if (new_config == nullptr) {
+      if ((new_config = provider->onConfigProtoUpdate(config_proto)) == nullptr) {
+        return false;
+      }
+    }
+    provider->onConfigUpdate(new_config);
+  }
+
+  return true;
+}
+
+void ConfigSubscriptionInstanceBase::bindConfigProvider(MutableConfigProviderImplBase* provider) {
+  // All config providers bound to a ConfigSubscriptionInstanceBase must be of the same concrete
+  // type; this is assumed by checkAndApplyConfig() and is verified by the assertion below. NOTE: a
+  // regular ASSERT() triggers a potentially evaluated expression warning from clang due to the args
+  // passed to the second typeid() call.
+  ASSERT_IGNORE_POTENTIALLY_EVALUATED(mutable_config_providers_.empty() ||
+                                      typeid(*provider) ==
+                                          typeid(**mutable_config_providers_.begin()));
+  mutable_config_providers_.insert(provider);
+}
+
+ConfigProviderManagerImplBase::ConfigProviderManagerImplBase(Server::Admin& admin,
+                                                             const std::string& config_name) {
+  config_tracker_entry_ =
+      admin.getConfigTracker().add(config_name, [this] { return dumpConfigs(); });
+  // ConfigTracker keys must be unique. We are asserting that no one has stolen the key
+  // from us, since the returned entry will be nullptr if the key already exists.
+  RELEASE_ASSERT(config_tracker_entry_, "");
+}
+
+const ConfigProviderManagerImplBase::ConfigProviderSet&
+ConfigProviderManagerImplBase::immutableConfigProviders(ConfigProviderInstanceType type) const {
+  static ConfigProviderSet empty_set;
+  ConfigProviderMap::const_iterator it;
+  if ((it = immutable_config_providers_map_.find(type)) == immutable_config_providers_map_.end()) {
+    return empty_set;
+  }
+
+  return *it->second;
+}
+
+void ConfigProviderManagerImplBase::bindImmutableConfigProvider(
```

Comment:
FWIW, the new naming scheme is also a bit confusing, since the config object that is provided is always immutable, but the fact that a new config can be swapped in is being assigned the mutability characteristic. IDK if it's worth bike shedding further on this, but just raising this as a potential reader confusion.